### PR TITLE
[koa] Allow to type response body

### PR DESCRIPTION
**Pull request by pkuczynski**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <https://github.com/koajs/koa/issues/1516>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

I was not able to type the response in any way, because `body` was defined as `any`. So I changed this to `unknown`, so it can be easily overridden by the type intersection.

Then I extended a Middleware definition to be able to pass the response body as 3rd type parameter.

Additionally, I provided a new type `Controller` which is similar to `Middleware`, but receiving only `Context` and returning `void`. Have a look at the tests to see example usage.

I am open for feedback @DavidCai1993 @jkeylu @brikou @harryparkdotio @chatoo2412

### CODE REVIEW

File: types/koa/index.d.ts
Comment (pkuczynski) [inline types/koa/index.d.ts:443]: Not sure where `CustomT` was coming from, so I gave it a more descriptive name

File: types/koa/index.d.ts
Comment (pkuczynski) [inline types/koa/index.d.ts:318]: This is the clue for the whole change. With only this small change, one could type the response like this:

```js
interface MyControllerResponse {
   ...
}

export const myController: Middleware<DefaultState, { body: MyControllerResponse, response: { body: MyControllerResponse } }> = async (ctx) => {
    ...
}
```

It was pretty long though, so I extended the Middleware type below...

File: types/koa/index.d.ts
Comment (harryparkdotio) [inline types/koa/index.d.ts:542]: I don't think it makes sense for this to become part of the Koa library types, within Koa, there isn't any reference to a `Controller`.

File: types/koa/index.d.ts
Comment (harryparkdotio) [inline types/koa/index.d.ts:721]: overriding these properties opens up the potential for being inconsistent, since these don't necessarily need to correlate with the existing reference to `body` and `response.body`.

File: types/koa/tsconfig.json
Comment (pkuczynski) [inline types/koa/tsconfig.json:15]: Old tests were failing with this setting. Not sure if this is important?

**Review by harryparkdotio — COMMENTED** [review]
Interested in hearing what other reviewers have to say, but I feel that this doesn't necessarily belong within the context of the Koa library types.

My thoughts are that you can override the types similar to how koa-bodyparser does ([ref](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/koa-bodyparser/index.d.ts))

File: types/koa/index.d.ts
Comment (pkuczynski) [reply types/koa/index.d.ts:542]: I was wondering about the same. Indeed Koa is not using this terminology. Yet found it pretty useful, as most of the people will need that, but maybe you right it belongs more to the @koa/router...

I am totally fine reverting this change :)

File: types/koa/index.d.ts
Comment (pkuczynski) [reply types/koa/index.d.ts:721]: There is a chance a middleware could modify the body, that's true, yet one could provide a type for what body could be populated before with... It's optional for the developer to define strict type here...

**Comment by pkuczynski** [comment]
The tests are failing on some other `koa*` packages. Not sure if I should fix them in this PR?

**Comment by pkuczynski** [comment]
@DavidCai1993 @jkeylu @brikou @chatoo2412 can you please have a look at this PR?

File: types/koa/index.d.ts
Comment (pkuczynski) [reply types/koa/index.d.ts:542]: Removed `Controller`

**Comment by pkuczynski** [comment]
@harryparkdotio I removed `Controller` stuff. Can you have another look and approve this PR?

File: types/koa/index.d.ts
Comment (harryparkdotio) [reply types/koa/index.d.ts:718]: I still don't find this to be appropriate to exist within this types package, as far as Koa is concerned, the response body can and should still be any.

This type enforcement can be handled within your application by extending the Koa types already

Comment (pkuczynski) [reply types/koa/index.d.ts:718]: That's correct. And so far with this PR it's still `any` by default, but allows developer to provide more specific type if he wants. Check line [537](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/50249/files#diff-bb94074d9ae1164d531dd6eb17a8858f06f4129662469a7d5e1dd19cb5eed86cR537)

```
type Middleware<StateT = DefaultState, ContextT = DefaultContext, ResponseBodyT = any> = compose.Middleware<
```

With this set-up I would be able to define my middleware as:

```
interface MyMiddlewareResponse {
  prop: string
}

export const myMiddleware: Middleware<{}, {}, MyMiddlewareResponse> = (ctx, next) => { ... }
```

Similarly like you can provide a more specific type for state and context. Is there anything wrong with it? I don't think so. Please tell me exactly what it would be? If you don't specify the third param for the `Middleware` it will be `any` by default, so this PR does not change current behaviour at all.

If there is another way to achieve the example I have shown above, meaning easily provide a type for response body, please could you provide an example?

Thank you in advance for your time :)

File: types/koa/tsconfig.json
Comment (sheetalkamat) [inline types/koa/tsconfig.json:14]: This cannot be turned off. You need to fix error instead

**Review by sheetalkamat — DISMISSED** [review]

File: types/koa/tsconfig.json
Comment (pkuczynski) [reply types/koa/tsconfig.json]: Of course. Sorry. Fixed.

**Comment by pkuczynski** [comment]
@DavidCai1993 @jkeylu @brikou @harryparkdotio @chatoo2412 @sheetalkamat anybody can approve this PR or suggest improvements?

**Comment by pkuczynski** [comment]
@tellnes would u mind having a look at this PR?

**Review by tellnes — APPROVED** [review]
This looks good to me. I'm looking forward to get this merged.

Btw, I would have split this into multiple pull requests since it does multiple things. Then it would have been easier to review and some of them could have merged while waiting for the approval of the stuff that needed discussion.

**Comment by pkuczynski** [comment]
Thanks for the feedback @tellnes. Next time I will fire up a few independent PRs...

**Comment by pkuczynski** [comment]
@sheetalkamat could you please re-review, as I fixed the issue which you reported...

**Review by orta — APPROVED** [review]
Nice, looks good to me - the 2.3 -> 3.0 is a breaking change but it still falls outside of what we considered ["supported"](https://github.com/DefinitelyTyped/DefinitelyTyped#older-versions-of-typescript-33-and-earlier) on DT, so I think this is fine 👍🏻

**Comment by pkuczynski** [comment]
Ready to merge

**Comment by cduff** [comment]
This looks to have caused the following issue? https://github.com/DefinitelyTyped/DefinitelyTyped/issues/51324

**Comment by rnbrady** [comment]
Seeing the following errors post upgrade:

```
error TS2339: Property 'body' does not exist on type 'Request'
error TS2345: Argument of type 'ParameterizedContext<any, IRouterParamContext<any, {}>>' is not assignable to parameter of type 'ParameterizedContext<DefaultState, DefaultContext, unknown>'.
```

I'm using:

```
koa@2.13.1
@types/koa@2.13.1
typescript@4.1.3
```

Problem could be my side but thought I'd let you know.

**Comment by pkuczynski** [comment]
@rnbrady do I read you correctly, that we should ignore your comment?

**Comment by rnbrady** [comment]
@pkuczynski yes, please ignore.

**Comment by xdumaine** [comment]
I am also getting `Property 'body' does not exist on type 'Request'` @rnbrady wast here a particular fix you made to get the request body?

**Comment by rnbrady** [comment]
@xdumaine apologies I don't recall any specifics from this issue. Good luck!

**Pull request merged**
